infoblox-ipam.go

```diff
@@ -214,22 +216,27 @@ func (ibDrv *InfobloxDriver) requestSpecificNetwork(netview string, subnet strin
 	return network, err
 }
 
-func (ibDrv *InfobloxDriver) RequestNetwork(netconf NetConfig, netviewName string) (network string, err error) {
+func (ibDrv *InfobloxDriver) RequestNetwork(netconf NetConfig, args *ExtCmdArgs, netviewName string, mac string) (network string, err error) {
 	var ibNetwork *ibclient.Network
+	var ibNetworks *[]ibclient.Network
 	// netviewName := netconf.IPAM.NetworkView
 	cidr := net.IPNet{}
 	log.Printf("RequestNetwork: IPAM.Subnet='%s'", netconf.IPAM.Subnet)
 	if netconf.IPAM.Subnet.IP != nil {
 		cidr = net.IPNet{IP: netconf.IPAM.Subnet.IP, Mask: netconf.IPAM.Subnet.Mask}
 		ibNetwork, err = ibDrv.requestSpecificNetwork(netviewName, cidr.String(), netconf.Name)
 	} else {
-		networkByName, err := ibDrv.objMgr.GetNetwork(netviewName, "", ibclient.EA{"Network Name": netconf.Name})
+		//networkByName, err := ibDrv.objMgr.GetNetwork(netviewName, "", ibclient.EA{"Network Name": netconf.Name})
+		networkByName, err := ibDrv.objMgr.GetNetworks(netviewName, "", ibclient.EA{"Network Name": netconf.Name})
```

Comment:
At least in the case of k8s, the above should not happen. 

Basically, if no CIDR is specified, then we have to assume that one flat IP space is used across all nodes. In which case, we have to ensure there is only ONE network with that network name. All IPs will be allocated in that network.

If CIDR is specified, then we must honor that CIDR. Here, we assume the CNI conf on each node has a unique CIDR specified. In this case, the network name EA will just be populated for informational purpose, but not searched.

The following are the k8s configurations for the about use cases. **These must be clearly documented**.


In k8s, we can only support the following two types of IPAM configuration:

- **One flat IP space across all nodes**

All CNI conf for the network must be identical across all nodes. For example

*One flat network 10.0.0.0/24*
>{
    "name": "ipam-test",
    "type": "macvlan",
    "master":"eth0",
    "ipam": {
         "type": "infoblox",
         "subnet": "10.0.0.0/24",
         "gateway":"10.0.0.1",
         "network-view": "cni_view"
     }
}

or

*One flat network automatically allocated by IB*
>{
    "name": "ipam-test",
    "type": "macvlan",
    "master":"eth0",
    "ipam": {
         "type": "infoblox",
         "network-view": "cni_view"
     }
}

- **Unique CIDR on each node**

In this case, a unique CIDR must be specified in the CNI conf on each node. This a typical use case if the nodes are auto-provisioned. In which case, during node provision time, IB is queried to allocate a subnet and the CNI conf generated accordingly. For example:

*Node 1*
>{
    "name": "ipam-test",
    "type": "macvlan",
    "master":"eth0",
    "ipam": {
         "type": "infoblox",
         "subnet": "10.0.0.0/24",
         "gateway":"10.0.0.1",
         "network-view": "cni_view"
     }
}

*Node 2*
>{
    "name": "ipam-test",
    "type": "macvlan",
    "master":"eth0",
    "ipam": {
         "type": "infoblox",
         "subnet": "10.0.1.0/24",
         "gateway":"10.0.1.1",
         "network-view": "cni_view"
     }
}

Reply:
> Basically, if no CIDR is specified, then we have to assume that one flat IP space is used across all nodes. In which case, we have to ensure there is only ONE network with that network name. All IPs will be allocated in that network.

> --network-container string
	Subnets will be allocated from this container if subnet is not specified in network config file (default "172.18.0.0/16") . To have multiple subnet add comma separated subnet. (ex. "192.168.0.0/24,192.169.0.0/24")
--prefix-length integer
	The CIDR prefix length when allocating a subnet from Network Container (default 24)

Then in that case this patch itself not needed. In this patch basically we are trying to Leverage Network container functionality and create sub network based on the prefix length below is the arg passed to IB. Looks like that doesn't hold good for K8s.

> Unique CIDR on each node

Current code is using the 'Network Name' EA for search as well. So we have to option 

1.  Either change the code
2.  Document and say user have to give the different name in each node.

Reply:
> Network Container

Network container functionality is not very useful for IP allocation for pods. BUT, it is definitely useful for creating unique CIDR for each node during auto-provisioning of nodes. Basically, in that case, you setup, for example, a /16 network container and allocate a unique /24 subnet out of that for each node.

> Current code is using the 'Network Name' EA for search as well

What I wanted to point out in my original comment is that, the code should NOT expect multiple networks returned when searching for Network Name EA (not that you should NEVER search for Network Name EA). Let me explain...
In the case of **One flat IP space across all nodes**, where no CIDR is specified in CNI conf, then the code would use Network Name EA for search. BUT, the code should only expect one network returned.
In the case of **Unique CIDR on each node**, because CIDR is specified in CNI conf, then the code should just use that CIDR, and there's no need to search using Network Name EA. 

> Document and say user have to give the different name in each node.

NO, I am pretty sure the Network Name has to be the same in all the nodes for the same overlay network. (unless I have misunderstood you)